### DISCUSSION TITLE
Plea for suggestion: speed up start of monitor mode when ignoring directories.

### DISCUSSION THREAD

**Discussion by Rmano** [abraunegg/onedrive#1966] [Q&A] [answered]
First of all: I *know* this is is a combined problem of my setup and OneDrive protocol, not an `ondedrive` issue! I am looking for suggestions...

So, I have a OneDrive account from my employer (and only one). The top directory (apart from all the cruft added by Office365) is like this:
```
root
|--- Backups
      |--- Live
|--- SharedFiles
|--- MySync
|--- ...
```
Into the `Backups` directory, there is a `Live` directory where a `rclone`-based process does a copy of (most of my) directories every evening, at the end of the workday. That thing is quite huge --- 92Gbyte of stuff. 

In my config for `onedrive` I have:
```
skip_dir = "Attic|Notebooks|Backups|Blocs de Notas"
monitor_interval = "60"
sync_business_shared_folders = "true"
```
It works ok, with just one problem: when I boot my machine at home, `onedrive` starts, and then it spends a lot of time (half an hour or something like that) receiving updates for the `Backup` folder and duly discarding them. That means that if I edit a file in `MySync` (or the file needs to be updated) I have to wait for the first `Sync with OneDrive is complete` to start working correctly.

Has anybody a suggestion about it? I know, the best thing would be to move the *#4$@ backup directory away, but I have just one account for the backup, and they are required (sort of) by my university, and they have to stay there on the corporate OneDrive unit...

Thank you! (and to @abraunegg , thanks^3 for this software).

**Comment by abraunegg** [comment]
@Rmano 
A couple of items to look at:
* Make sure you are using v2.4.17 or later
* Do not use a monitor interval of 60 .... you are compounding your problem by doing this
* Set your full scan frequency appropriatly - please read https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#monitor_fullscan_frequency

The other aspect would be a feature request to 100% ignore the integrity check at startup .... but it is there for a reason

Also - using rclone will also change all the /delta data points ... so using rclone will also be probably impacting you

**Reply by Rmano** [reply]
Thanks for answering! 

Yes, I am using `v2.4.17`. 

About the monitor interval... ¿it's too short? I will try to reset it to 300, and I will try to bump up a bit the full scan frequency. I agree that skipping integrity checks seems not a good idea... (the ideal thing would be to have a way to say the OneDrive server to completely ignore a subtree, but I suspect that this does not exist in the protocol).

I did suspect that `rclone` could create the problem... but I do not know what to do exactly to replace it. Basically, now I am doing a 

```
rclone sync --exclude-from="$rclonedir/filtering.txt"\
        --fast-list --skip-links\
        -P --log-file="$rclonedir/$mydir".log --log-level=INFO\
        "$mydir" comillasOD:Backups/Live/"$mydir"
```

for each `$mydir` that I have to backup.

**Reply by Rmano** [reply]
> The other aspect would be a feature request to 100% ignore the integrity check at startup .... but it is there for a reason

Wait, if I understand correctly, the check is done periodically also? Maybe it could be *deferred* or made in parallel with the other operations... but I am just thinking aloud, I have too little idea about the protocol.

**Comment by abraunegg** [comment, marked as answer]
@Rmano 
> About the monitor interval... ¿it's too short?

Yes .. 60 seconds is *way* too short ... this should be aligned to your 'rate-of-data-change' online ... if your data online is changing every 5 mins, make it 5 mins (300 seconds) .. if your data is not changing often - once every few hours, every 12 hours .. adjust accordingly.

Do not use a value of 60 ... 

> I did suspect that rclone could create the problem... but I do not know what to do exactly to replace it.

Please read the [documentation](https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md):
* `--single-directory`
* `--upload-only` if required
* `--download-only` if required

> Wait, if I understand correctly, the check is done periodically also? Maybe it could be deferred

Yes .. governed by the value you have set to 60 seconds + monitor_fullscan_frequency.

The whole point of doing it at startup is so that your client is in a consistent state ... the last thing you want is your data to be inconsistent.

> Maybe it could be deferred or made in parallel with the other operations

Due to how OneDrive API works + how the client operates this is not possible. It *has* to be done at the start as this is the *most* efficient point to do this.

**Comment by Rmano** [comment]
Thanks again for spending your time to help!

>  if your data online is changing every 5 mins, make it 5 mins (300 seconds) .. if your data is not changing often - once every few hours, every 12 hours .. adjust accordingly. 

Yep, that is the problem. Part of the directories are for collaborative, online work. So in that case even 5 minutes is a long time --- when we are adjusting a collection of files (say, 50+ word documents with the contents of our courses) and a couple of summary files, for example. The backup directory changes only overnight, though...
 
>  `--single-directory`

Thanks, I'll read it, I am not sure it'll work. The `Backup` directory contains files that are *outside* my `~/OneDrive` directory --- for example, my `~/research` hierarchy. I just need a kind of `rsync`... this is why I am using `rclone` now. Anyway, I'll explore that. Can I use `--single-directory` while another `onedrive` is running in the background in monitor mode?

Romano

**Comment by abraunegg** [comment]
@Rmano 
Why cant you run 2 instances of the client, each with their own configuration?

* Client 1 - excludes the backup directory
* Client 2 - includes *only* the backup directory

This way you can have different sync timings and operational aspects for the content data as is appropriate

**Reply by Rmano** [reply]
Yes, I'll try this. The problem is that the OneDrive root has both directories in its root (It would be solved with two OneDrive accounts, probably), so the fast-paced client would still have to do the integrity check on the OneDrive/Backup directory and so the problem of having to wait for the (two hours, bad network this morning) startup time after a nightly backup would be still there. 

I will mark your previous answer as the answer, as I think that this is the best I can do given the protocol (and I'll try to move the backup elsewhere if they let me do it). 

Maybe I should give a try at https://github.com/jstaf/onedriver --- could be more fitting for my usage, especially if I can find a way to blacklist remote directories (I found the link in one of your answers here!)

**Reply by Rmano** [reply]
Hmmm, no, `onedriver` will not really solve my problems due to https://github.com/jstaf/onedriver/issues/252#issuecomment-1111713311

**Reply by abraunegg** [reply]
> so the fast-paced client would still have to do the integrity check on the OneDrive/Backup directory and so the problem of having to wait for the (two hours, bad network this morning) startup time after a nightly backup would be still there.

No ... if it is *excluded* even if the data is there, the data should not be integrity checked .. as it should be skipped / bypassed.

If this is not occurring .. then that is a bug

**Comment by abraunegg** [comment]
@Rmano 
You can have everything in a single OneDrive root .. and use the same account to access ... 

With separate configurations running, each would be doing its own aspect - and, on your more realtime data, data that is to be excluded is excluded and not checked ....

My suggestion is to give that a shot

**Reply by Rmano** [reply]
Yes, I'll try it. What I doubt is --- if I could tell the `onedrive` client that I have now to not check the `Backup` subdirectory that will suffice --- I mean (maybe I'm dense --- Monday morning, you know) what will be the difference? I am just excluding it now, but nevertheless, I receive the stream of changes for it during the consistency check on startup...

**Reply by abraunegg** [reply]
>  if I could tell the onedrive client that I have now to not check the Backup subdirectory that will suffice

In 'client config 1' exclude the required directories via 'skip_dir' configuration. Potentially even use 'sync_list' to specify all directories and exclude the required path as well. Use `onedrive --confdir '~/.config/onedrive-realtime/' <plus any other options>`

In 'client config 2' exclude all directories except for 'Backup' via 'sync_list' configuration. Use `onedrive --confdir '~/.config/onedrive-backup/' <plus any other options>`

In this scenario - client 1, is doing your 'real-time' work - 60 second check, potentially with 'monitor_fullscan_frequency' set to 720 (12 hrs) or even 1080 (18 hrs). This client should be skipping and ignoring any change or data in 'Backup'

Client 2 would be using the 'monitor_interval' = 21600 (6 hrs) and 'monitor_fullscan_frequency' set to 4 (once per day) - this means that the Backup folder is only being scanned and integrity checked once per day, and, as your data for this is only changing on an infrequent basis - this is what seems ideal.

**Comment by Rmano** [comment]
Ok, I think I got it. But now I have this in my config:

```
$ cat config | grep -v -E "^#" 
skip_file = "~*|.~*|*.tmp|Romano @ Universidad*"
skip_dir = "Attic|Notebooks|Backups|Blocs de Notas"
monitor_fullscan_frequency = "24"
sync_business_shared_folders = "true"
```

So the directory `Backups` should be skipped? In the verbose log, I have *thousands* of these: 
```
[DEBUG] Processing change 2 of 235
  [DEBUG] OneDrive Change: {"@odata.etag":"\"{9C6D8D12-C507-425B-AF08-85644659DC4B},1\"","@odata.type":"#microsoft.graph.driveItem","cTag":"\"c:{9C6  D8D12-C507-425B-AF659DC4B},0\"","eTag":"\"{9C6D8D12-C507-425B-AF08-85644659DC4B},1\"","fileSystemInfo":{"createdDateTime":"2020-11-19T09:5  1:09Z","lastModifiedDateTime":"2020-11-19T09:51:09Z"},"folder":{"childCount":11},"id":"01CRET2RY6FLNBK6CEFMRDFTXCL","name":"Backups","paren  tReference":{"driveId":"b!LtHke7bh3kG0bVzBE1QkHnhNqI5o-VcAsHJ_LJBMtZ3mQagpnag-F9Cq","driveType":"business","id":"01CRET2R56725BZO35  4PWSELRRZ","path":"\/drives\/b!LtHke7bh3kG0bVzBE1QCWW-VcAsHJ_LJBMtZ3mQagpnag-F9Cq\/root:"},"size":152863209297}
``` 

...referring to the `Backups` subdirectory. In the `/delta` API answer there are a lot of references to the excluded dir. ¿Do I have any problem with my config?

To be clear: I do *not* want to have a copy of my `Backups` directory in the local `~/OneDrive` directory, and I do not have it with that configuration, which is ok and as expected. The problem is that at the first connection, there are thousands of delta information bits from the server that are correctly ignored, but the thing go on and on during a lot of time, before starting to work correctly. I think that this is really unavoidable... because my configuration is what `client config 1` you suggested should be, and I do not need the `client config 2`...


PD is it possible to have access to my account with the content of the log? Otherwise, I have no problem to share it (no real secrets and I trust you if needed to send to an email).

**Reply by abraunegg** [reply]
> So the directory `Backups` should be skipped? In the verbose log, I have _thousands_ of these:
> 
> ```
> [DEBUG] Processing change 2 of 235
>   [DEBUG] OneDrive Change: {"@odata.etag":"\"{9C6D8D12-C507-425B-AF08-85644659DC4B},1\"","@odata.type":"#microsoft.graph.driveItem","cTag":"\"c:{9C6  D8D12-C507-425B-AF659DC4B},0\"","eTag":"\"{9C6D8D12-C507-425B-AF08-85644659DC4B},1\"","fileSystemInfo":{"createdDateTime":"2020-11-19T09:5  1:09Z","lastModifiedDateTime":"2020-11-19T09:51:09Z"},"folder":{"childCount":11},"id":"01CRET2RY6FLNBK6CEFMRDFTXCL","name":"Backups","paren  tReference":{"driveId":"b!LtHke7bh3kG0bVzBE1QkHnhNqI5o-VcAsHJ_LJBMtZ3mQagpnag-F9Cq","driveType":"business","id":"01CRET2R56725BZO35  4PWSELRRZ","path":"\/drives\/b!LtHke7bh3kG0bVzBE1QCWW-VcAsHJ_LJBMtZ3mQagpnag-F9Cq\/root:"},"size":152863209297}
> ```
> 
> ...referring to the `Backups` subdirectory. In the `/delta` API answer there are a lot of references to the excluded dir. ¿Do I have any problem with my config?

That is 100% normal ... why?

Because the OneDrive API has no knowledge of 'client side rules' (which skip_dir et al are) - thus - sends you all changes from online - which need to be processed against your client side rules - this is why they are sent.

Now .. the change sent, if it *matches* a 'skip_dir' rule - it should spit out, that the change is being skipped ... ie .. not processed.

> PD is it possible to have access to my account with the content of the log? Otherwise, I have no problem to share it (no real secrets and I trust you if needed to send to an email).

Not possible to access the account via the log .. unless you also start sharing out your API secret as well .. then you have a bigger problem.

**Reply by Rmano** [reply]
Ok, I'll try a debug sync and to share the result with you. But yes, the things are skipped, at the end of the log entry I have: 

```
[DEBUG] 'item id' matches search 'id'                        = false
  [DEBUG] 'parentReference id' matches search 'id'             = false
  [DEBUG] 'thisItemParentPath' contains 'syncFolderChildPath'  = true
  [DEBUG] 'thisItemParentPath' contains search 'id'            = false
  [DEBUG] Change matches search criteria to apply
  [DEBUG] Flagging as unwanted: find(item.parentId).length != 0
  [DEBUG] Skipping OneDrive change as this is determined to be unwanted
  [DEBUG] ------------------------------------------------------------------
``` 

so I think it's working as it should. The problem is that after a backup, sometimes there are thousands (or more) of that entries, and that means half an hour (or more) delay (with `onedrive` at almost 100% CPU) before it becomes sensitive to changes in the "not ignored" directories. This is why I think this is unavoidable... I have to try to move the Backup directory away.

I really appreciate your help, BTW; I feel a bit dumb sometimes. Thanks.

**Reply by abraunegg** [reply]
> To be clear: I do _not_ want to have a copy of my `Backups` directory in the local `~/OneDrive` directory, and I do not have it with that configuration, which is ok and as expected. The problem is that at the first connection, there are thousands of delta information bits from the server that are correctly ignored, but the thing go on and on during a lot of time, before starting to work correctly.

On the *very first client startup* this should be true - on subsequent startups - only the delta changes should be sent.

Are you forcing or using `--resync` somehow ... of you are - you are then sending *all* of your file details from OneDrive to re-create the local state ... do not use `--resync` unless the application specifically advises to.

**Reply by Rmano** [reply]
BTW, I can send the log somewhere to you? When it happens again? There are no secrets but some details could be private (names of people etc in file names) so I prefer not to publish it.

**Reply by abraunegg** [reply]
> I really appreciate your help, BTW; I feel a bit dumb sometimes. Thanks.

Happy to even do a conference call / screen share / real-time debug with you if that also helps sort out what is going on.

**Reply by abraunegg** [reply]
> BTW, I can send the log somewhere to you? When it happens again? There are no secrets but some details could be private (names of people etc in file names) so I prefer not to publish it.

Have a read of https://github.com/abraunegg/onedrive#reporting-an-issue-or-bug 

Drop an email to the address - and I will give you my contact details.

**Reply by Rmano** [reply]
> On the very first client startup this should be true - on subsequent startups - only the delta changes should be sent. 

Hmmm. I suppose "very first" is not after a reboot, but the first time it starts (and that the status is persistent somehow). The other option is that the `rclone` is really messing up things marking everything changed even if it is not. Some sort of access time... (shouldn't do, but I'll check). 

Thank you for the offer to debug, it's way more than I expected. I will try to see what happens tomorrow (or after the next nightly backup when it happens that I am remote working).
